plugins/linux_log_parser: Add support for test name regex

katieworton opened the pull request:
Add support for a test name regex to extract parts of the log parser outputs to produce more meaningful test names.

Instead of producing test names with the "REGEX_NAME" + SHA, create test names with "REGEX_NAME"+"REGEX_EXTRACT_NAME"+SHA to make the test names easier to interpret.

Example test name before:
check-kernel-oops-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220

Example test name after:
check-kernel-oops-oops-bug-preempt-smp-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220

katieworton committed:
plugins/linux_log_parser: Add support for test name regex

Add support for a test name regex to extract parts of the log parser
outputs to produce more meaningful test names.

Instead of producing test names with the "REGEX_NAME" + SHA, create test
names with "REGEX_NAME"+"REGEX_EXTRACT_NAME"+SHA to make the test names
easier to interpret.

Example test name before:
check-kernel-oops-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220

Example test name after:
check-kernel-oops-oops-bug-preempt-smp-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton commented on test/plugins/test_linux_log_parser.py:
```diff
@@ -259,7 +259,7 @@ def test_sha_name(self):
 
         # Now check if a test with sha digest in the name
-        test = testrun.tests.get(suite__slug='log-parser-boot', metadata__name='check-kernel-oops-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220')
+        test = testrun.tests.get(suite__slug='log-parser-boot', metadata__name='check-kernel-oops-oops-bug-preempt-smp-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220')
```
We could do this. This would mean the tests created would look like:
- check-kernel-oops-bug-preempt-smp
- check-kernel-oops-bug-preempt-smp-<sha>
- check-kernel-oops-bug-preempt-smp-<sha>
- ...etc
- check-kernel-<name2>
- check-kernel-<name2>-<sha>
- check-kernel-<name2>-<sha>
- ...etc

Instead of:
- check-kernel-oops
- check-kernel-oops-bug-preempt-smp-<sha>
- check-kernel-oops-bug-preempt-smp-<sha>
- ...etc
- check-kernel-<name2>-<sha>
- check-kernel-<name2>-<sha>
- ...etc

So there would be more tests generated but each "big bucket" (before dividing by sha) would have less in it.

Is change helpful?

roxell replied:
yes, make sense and its super helpful =)

katieworton replied:
I have pushed an update to move the extracted name regex to the higher-level log parser bucket as discussed here.

The name is still like "check-kernel-oops-oops-bug-preempt-smp" rather than "check-kernel-oops-bug-preempt-smp" since I didn't want to risk losing the "oops" part (or equivalent) in case the naming regex fails to extract anything (so we can fall back on "check-kernel-oops" in that case).

Let me know what you think :)

katieworton commented:
@chaws @terceiro @nareshkamboju @roxell - If you can look over this PR and highlight if there are any issues with the proposed naming updates that would be great.

To keep this work on track, the PR should be merged this week - so if there are any obvious issues it would be great if you could highlight these ASAP.

Thanks!

chaws commented on squad/plugins/linux_log_parser.py:
```diff
@@ -4,26 +4,28 @@
-    ('check-kernel-exception', r'-+\[? cut here \]?-+.*?-+\[? end trace \w* \]?-+'),
-    ('check-kernel-kasan', r'=+\n\[[\s\.\d]+\]\s+BUG: KASAN:.*?=+'),
-    ('check-kernel-kfence', r'=+\n\[[\s\.\d]+\]\s+BUG: KFENCE:.*?=+'),
+    ('check-kernel-exception', r'-+\[? cut here \]?-+.*?-+\[? end trace \w* \]?-+', r"\d][^\+\n]*"),
```
I saw you added some text to other patterns, like `BUG: KASAN:`, `Unhandled` etc, but for check-kernel-exception there's no text.

katieworton replied:
Thanks! I didn't add it to this case since I didn't think including "cut here" in the name was very useful. There was nothing else guaranteed to always be in the extracted log line for this case.

chaws replied:
humm, makes sense. I think the real test name will come from its contents anyways. All good!

chaws commented on squad/plugins/linux_log_parser.py:
```diff
@@ -106,11 +111,27 @@ def __create_tests(self, testrun, suite, test_name, lines):
+        snippet = matches[0]
+        without_numbers_and_time = self.__remove_numbers_and_time(snippet)
+
+        return slugify(without_numbers_and_time)[:191]
```
Why `191` here?

katieworton replied:
Updated with a code comment - max name length is 256 for the name in SQUAD and we have 64 character SHA + a "-" to accommodate

chaws commented:
> @chaws @terceiro @nareshkamboju @roxell - If you can look over this PR and highlight if there are any issues with the proposed naming updates that would be great.
> 
> To keep this work on track, the PR should be merged this week - so if there are any obvious issues it would be great if you could highlight these ASAP.
> 
> Thanks!

It is really well done :), but I found a couple of issues I commented in line. Also, please add a few tests to make sure the created naming matches the expected. Just like you did for test `check-kernel-oops-oops-bug-preempt-smp-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220`

katieworton committed:
plugins/linux_log_parser: Move test name regex to higher level

Move the meaningful name extracted by the REGEX_EXTRACT_NAME regex to be
included in the "higher-level" test name.

Before, test names would look like:
"check-kernel-oops"
And these would then be broken down into smaller test buckets with names
like:
"check-kernel-oops-oops-preempt-smp-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220"

Now, the higher-level test would be called:
"check-kernel-oops-oops-preempt-smp"
And the broken down tests would still look like:
"check-kernel-oops-oops-preempt-smp-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220"

This makes the higher-level test names more meaningful, then if there
are differences which are not captured by the higher-level name, the SHA
can be used to differentiate the different cases.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton commented:
> > @chaws @terceiro @nareshkamboju @roxell - If you can look over this PR and highlight if there are any issues with the proposed naming updates that would be great.
> > To keep this work on track, the PR should be merged this week - so if there are any obvious issues it would be great if you could highlight these ASAP.
> > Thanks!
> 
> It is really well done :), but I found a couple of issues I commented in line. Also, please add a few tests to make sure the created naming matches the expected. Just like you did for test `check-kernel-oops-oops-bug-preempt-smp-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220`

Thank you, @chaws!

I've made some updates based on the comment from @roxell: I moved the regex name up a level. So test called `check-kernel-oops` would now be called `check-kernel-oops-oops-bug-preempt-smp` (the version with the SHA would still be `check-kernel-oops-oops-bug-preempt-smp-a1acf2f0467782c9c2f6aeadb1d1d3cec136642b13d7231824a66ef63ee62220`)

Happy to add more tests but let me know what you think of this change and we can go from there :)

katieworton committed:
test/plugins/test_linux_log_parser: Add an extra test for SHAs

Add an extra test case for multiple failures with SHAs.

Signed-off-by: Katie Worton <katie.worton@linaro.org>

katieworton commented:
@chaws - added some more unit tests for the SHA versions as I realised this was not dependent on the `check-kernel-oops` vs `check-kernel-oops-oops-bug-preempt-smp` change.

After speaking with @roxell, I also created a draft patch to try and change the naming from things like `check-kernel-oops-oops-bug-preempt-smp` to `check-kernel-oops-bug-preempt-smp` (removing the repetition): https://github.com/Linaro/squad/pull/1144

However, I had some issues with getting the unit tests going with https://github.com/Linaro/squad/pull/1144. I feel like we should just get this initial PR (https://github.com/Linaro/squad/pull/1142) merged then do another set of changes with improvements once we start seeing the outputs from SQUAD.

chaws commented:
> We also talked about removing "check-kernel-oops-" completely. @chaws is there any meaning of having "check-kernel-(oops|bug|warning|...) and having them "passing" if nothing if found? Can't we just drop the "passing" tests all together?

If I recall correctly, you were the one who asked for this :) so if you ask for removal let's just do it then

chaws reviewed (approved):
LGTM
